### ISSUE TITLE
Homematic IP WTH-2

### ISSUE THREAD

**Issue by vakilando** [danielperna84/pyhomematic#223]
I hope I'm not wrong here.....
Is it possible to add the **HmIP-WTH-2 (Wandthermostat)** device to the supported device list?

**My Settings**
I have a CCU2 (on raspi piVCCU) and HMIP-eTRV-2 at home. 
I use Home Assistant (on Unraid docker) to read and set temperature of those thermostats and this works fine!
Now I bought a HMIP WTH-2 (Wandthermostat) and linked it in the CCU2 with two HMIP-eTRV-2 in the same room. 
The HMIP WTH-2 and the linked HMIP-eTRV-2 work fine when using the CCU2 on piVCCU.
Sadly the The Homematic IP WTH-2 ("Wandthermostat") is not fully working in HA.

- Setting temperature in the HA GUI or in automations for WTH-2 seams to work as the WTH-2 display displays the new temperature. But the new temperature will not be set on the the linked thermostats.
- Setting the operation mode (auto/manual/boost) in the HA GUI or in automations for WTH-2 doesn't work
- When changing the operation mode on the WTH-2 itself you can see that it changes also in HA GUI
- Setting the operation mode or temperature of HMIP thermostat (HmIP-eTRV,-2) in the HA GUI or in automations works fine

**Describe the solution you'd like**
I would like that setting the operation mode (auto/manual/boost) and the temperature in the HA GUI or in automations for WTH-2 works as expected.

**Describe alternatives you've considered**
There is no real alternative... I could sell the WTH-2 again and set the temperature of each thermostat individually but the great benefit of WTH-2 is that I can set the temperature of all  thermostats in a room in dependency of the temperature measured by WTH-2.

**Error Logs of HA:**
```
Log Details (ERROR)
Sun Feb 10 2019 22:53:07 GMT+0100 (Mitteleuropäische Normalzeit)
Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/app/homeassistant/helpers/service.py", line 289, in _handle_service_platform_call
    await getattr(entity, func)(**data)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/app/homeassistant/components/climate/homematic.py", line 142, in set_operation_mode
    self._hmdevice.MODE = code
  File "/usr/local/lib/python3.6/site-packages/pyhomematic/devicetypes/thermostats.py", line 73, in MODE
    self.actionNodeData(mode, set_data)
  File "/usr/local/lib/python3.6/site-packages/pyhomematic/devicetypes/generic.py", line 345, in actionNodeData
    return self._setNodeData(name, self.ACTIONNODE, data, channel)
  File "/usr/local/lib/python3.6/site-packages/pyhomematic/devicetypes/generic.py", line 359, in _setNodeData
    (name, data, nodeChannel))
UnboundLocalError: local variable 'nodeChannel' referenced before assignment
```

Tell me what I can do to help implement the HMIP-WTH 2 to work with HA

**Comment by danielperna84** [comment]
1. What you describe with the temperature just being set on the single device and not the group sounds familiar. If I recall correctly that's a limitation of the XML-RPC API. _Internally_ your group is actually available as a separat _device_ which you can control. The relevant section of the documentation:

 > The virtual thermostatgroups the CCU provides use port 9292 and require you to set the path setting to /groups

I don't know if the port is different for HmIP-Groups. But please see what happens if you add that part to your configuration.

2. To get the operational modes integrated we will have to do some testing where you have to modify some code. That's hard to do with Docker setups. If it's possible, please follow [these](https://github.com/danielperna84/pyhomematic/pull/201#issuecomment-451025838) instructions until step 1. where it says to edit `actors.py`. In your case it will be `thermostats.py`. I will try to figure out the changes you will need to make there during this weekend and report back so you can test it.

**Comment by vakilando** [comment]
Ok, I did some testing, troubleshooting and searching in the www. 
I also found this post: [https://homematic-forum.de/forum/viewtopic.php?t=41706](https://homematic-forum.de/forum/viewtopic.php?t=41706). 
I have to admit that - after waiting longer for the change to take effect - the two HmIP eTRV-2 indeed change the temperature after setting it on the HMIP WTH-2. My problems are the same as described there.
But
- it takes 10 to 20 minutes after the HMIP WTH-2 was set that the first HmIP eTRV-2 gets the information and sets the temperature accordingly to the HMIP WTH-2!
- it takes another 10-20 min that the other HmIP eTRV-2 gets the information and sets the temperature...! 

So I think in HA I (sadly) cannot rely on the HMIP WTH-2 but must change the HMIP WTH-2 AND the two HmIP eTRV-2...

Adding "groups" to the HA configuration.yaml didn't change anything:
```yaml
homematic:
  interfaces:
    rf:
      host: xxx
      #resolvenames: json
      username: !secret xxx
      password: !secret xxx
    ip:
      host: xxx
      port: 2010
      #resolvenames: json
      # neu test
      username: !secret xxx
      password: !secret xxx
    groups:
      host: xxx
      port: 9292
      resolvenames: json
      username: !secret xxx
      password: !secret xxx
      path: /groups
  hosts:
    ccu2:
      host: xxx
      username: !secret xxx
      password: !secret xxx

climate:
  platform: homematic
```
The operation mode problem stays the same:
- Setting the operation mode (auto/manual/boost) in the HA GUI or in automations for WTH-2 doesn't work
- When changing the operation mode on the WTH-2 itself you can see that it changes also in HA GUI

I will now try to now to follow [these](https://github.com/danielperna84/pyhomematic/pull/201#issuecomment-451025838) instructions and come back to tell what happend....

**Comment by danielperna84** [comment]
> Adding "groups" to the HA configuration.yaml didn't change anything:

Also regarding the temperature? That's what I initially thought could fix the problem of setting the temperature and having all devices within the group react on that. 



> I will now try to now to follow [these](https://github.com/danielperna84/pyhomematic/pull/201#issuecomment-451025838) instructions and come back to tell what happend....

Nothing will change at the beginning. I will have to look at the device documentation and guess which channel is the correct one to set the mode. That information then needs to be added to `thermostats.py`. So if you follow the instructions and have a setup that's functionally identical to your productive Home Assistant (you only need to configure HomeMatic, the rest is irrelevant), then is it's time to modify the source code. 😉

**Comment by danielperna84** [comment]
I had a look now. In the `thermostats.py` you'll find a class called `IPThermostatWall`. Replace the whole class with the following code:

```python
class IPThermostatWall(HMThermostat, HelperLowBatIP):
    """
    HmIP-STHD
    ClimateControl-Wall Thermostat that measures temperature and allows to set a target temperature or use some automatic mode.
    """
    def __init__(self, device_description, proxy, resolveparamsets=False):
        super().__init__(device_description, proxy, resolveparamsets)

        # init metadata
        self.SENSORNODE.update({"ACTUAL_TEMPERATURE": [1],
                                "HUMIDITY": [1]})
        self.WRITENODE.update({"SET_POINT_TEMPERATURE": [1]})
        self.ACTIONNODE.update({"CONTROL_MODE": [1],
                                "BOOST_MODE": [1]})
        self.ATTRIBUTENODE.update({"LOW_BAT": [0],
                                   "SET_POINT_MODE": [1]})

    def get_set_temperature(self):
        """ Returns the current target temperature. """
        return self.getWriteData("SET_POINT_TEMPERATURE")

    def set_temperature(self, target_temperature):
        """ Set the target temperature. """
        try:
            target_temperature = float(target_temperature)
        except Exception as err:
            LOG.debug("Thermostat.set_temperature: Exception %s" % (err,))
            return False
        self.writeNodeData("SET_POINT_TEMPERATURE", target_temperature)

    @property
    def MODE(self):
        """ Return mode. """
        if self.getAttributeData("BOOST_MODE"):
            return self.BOOST_MODE
        else:
            return self.getAttributeData("SET_POINT_MODE")

    @MODE.setter
    def MODE(self, setmode):
        """ Set mode. """
        if setmode == self.BOOST_MODE:
            self.actionNodeData('BOOST_MODE', True)
        elif setmode in [self.AUTO_MODE, self.MANU_MODE]:
            if self.getAttributeData("BOOST_MODE"):
                self.actionNodeData('BOOST_MODE', False)
            self.actionNodeData('CONTROL_MODE', setmode)

    def turnoff(self):
        """ Turn off Thermostat. """
        self.writeNodeData("SET_POINT_TEMPERATURE", self.OFF_VALUE)
```
With that you should at least be able to turn on the `BOOST_MODE`. I don't know how that should be done in the UI though. Since Home Assistant has changed to the Lovelace UI I can only switch between manual and automatic mode. To do something else I probably have to use the `climate.set_operation_mode` service.

**Comment by vakilando** [comment]
I will test the modifications for the operation-mode tomorrow (hmmm today....).
I've just updated my piVCCU to piVCCU3 and restored my backup to be sure to have the latest firmware.


> Also regarding the temperature? That's what I initially thought could fix the problem of setting the temperature and having all devices within the group react on that.

Setting the temperature with the HMIP-WTH-2 over the HA Web-GUI works now! 
(perhaps after the piVCCU3 upgrade? I don't know...)

- Both HmIP eTRV-2 got the settings about 3 minutes after setting the HMIP-WTH-2.
But it's not shown in the HA GUI, only on the display of both HmIP eTRV-2.
- About another 10 minutes later I can see the changes of one HmIP eTRV-2 in the HA Web-GUI

I'm going to bed now and do further testing tomorrow. Thank you for your support!

**Comment by DieterChmod** [comment]
I'm working on a fix for that. Coming soon.

**Comment by DieterChmod** [comment]
Sorry, I was a bit in a hurry. Actually it is only one thing you should do.
Please add the new "IPThermostatWall2" class to HM_IGNORE_DISCOVERY_NODE_EXCEPTIONS in __init__.py in the homematic component of homeassistant. :)

**Comment by DieterChmod** [comment]
I know, this doesn't really belong here, but could you also add 'IPThermostat' to DISCOVER_SENSORS and HM_IGNORE_DISCOVERY_NODE_EXCEPTIONS?
Then the IP thermostats would be treated the same as the non-IP ones, i.e. there will be a an auto-detected sensor for temperature.

**Comment by danielperna84** [comment]
Fixes with [0.1.57](https://github.com/danielperna84/pyhomematic/releases/tag/0.1.57). I'll take care of the changes in Home Assistant.
